clang/runtime/dpct-rt/include/dpct/blas_utils.hpp

```diff
@@ -2467,8 +2467,618 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
+                     const void *a, matrix_layout_ptr a_desc, const void *b,
+                     matrix_layout_ptr b_desc, const void *beta, const void *c,
+                     matrix_layout_ptr c_desc, void *d,
+                     matrix_layout_ptr d_desc, dpct::queue_ptr q_ptr);
+};
+
+namespace detail {
+dnnl::memory::data_type dpct_type_to_dnnl_type(library_data_t type) {
+  switch (type) {
+  case library_data_t::real_half:
+    return dnnl::memory::data_type::f16;
+  case library_data_t::real_bfloat16:
+    return dnnl::memory::data_type::bf16;
+  case library_data_t::real_float:
+    return dnnl::memory::data_type::f32;
+  case library_data_t::real_double:
+    return dnnl::memory::data_type::f64;
+  case library_data_t::real_int32:
+    return dnnl::memory::data_type::s32;
+  case library_data_t::real_int8:
+    return dnnl::memory::data_type::s8;
+  case library_data_t::real_uint8:
+    return dnnl::memory::data_type::u8;
+  default:
+    throw std::runtime_error(
+        "the input library_data_t type has no corresponding type in dnnl");
+  }
+}
+} // namespace detail
+
+void matmul(matmul_desc_ptr compute_desc, const void *alpha, const void *a,
+            matrix_layout_ptr a_desc, const void *b, matrix_layout_ptr b_desc,
+            const void *beta, const void *c, matrix_layout_ptr c_desc, void *d,
+            matrix_layout_ptr d_desc, dpct::queue_ptr q_ptr) {
+  if (!q_ptr)
+    q_ptr = &get_default_queue();
+  if (compute_desc->_pointer_mode == pointer_mode_t::device_vector ||
+      compute_desc->_pointer_mode ==
+          pointer_mode_t::alpha_device_vector_beta_zero ||
+      compute_desc->_pointer_mode ==
+          pointer_mode_t::alpha_device_vector_beta_host) {
+    throw std::runtime_error("Unsupported pointer mode.");
+  }
+
+  // 1. if beta is not zero, need throw exception since we do not support it.
+  if (beta != nullptr) {
+    size_t beta_size =
+        dpct::detail::library_data_size[static_cast<unsigned int>(
+            compute_desc->_scale_type)] /
+        8;
+    void *beta_host = std::malloc(beta_size);
+    void *beta_zero = std::malloc(beta_size);
+    std::memset(beta_zero, 0, beta_size);
+    q_ptr->memcpy(beta_host, beta, beta_size).wait();
+    if (std::memcmp(beta_host, beta_zero, beta_size))
+      throw std::runtime_error("Non-zero beta is not supported.");
+  }
+
+  // 2. if has epilogue, need throw exception since we do not support it.
+  if (compute_desc->_epilogue != 1) {
+    throw std::runtime_error("Epilogue is not supported.");
+  }
+
+  // 3. only support row major and col major
+  if ((a_desc->_order != order_t::col && a_desc->_order != order_t::row) ||
+      (b_desc->_order != order_t::col && b_desc->_order != order_t::row) ||
+      (c_desc->_order != order_t::col && c_desc->_order != order_t::row) ||
+      (d_desc->_order != order_t::col && d_desc->_order != order_t::row)) {
+    throw std::runtime_error("Only support row majorand col major.");
```

Comment:
We can transform other layouts to row major layout (allocate new memory). After computation, transform it back to the original memory. Maybe also emit some runtime messages.

Reply:
Calculation impl done. Runtime message/warning message not added yet.

Reply:
done

Reply:
@abhilash1910,
For non-col-major layout input, I convert it to the col-major layout before computation.
For non-col-major layout output, I use the col-major layout to do the computation; after computation, I convert the result to the target layout.